container.go

```diff
@@ -0,0 +1,592 @@
+		} else if err != nil {
+			return &ContainerError{Container: container, Operation: operation, Err: err}
+		}
+
+		logrus.Debugf(title+" succeeded id=%s", container.id)
+		return nil
+	}
+
+	finished, err := waitTimeoutHelper(container, timeout)
+	if !finished {
+		return ErrTimeout
+	} else if err != nil {
+		return &ContainerError{Container: container, Operation: operation, Err: err}
+	}
+	logrus.Debugf(title+" succeeded id=%s", container.id)
+	return nil
+}
+
+func (container *container) hcsWait(timeout uint32) (bool, error) {
+	var (
+		resultp   *uint16
+		exitEvent syscall.Handle
+	)
+
+	err := hcsCreateComputeSystemWait(container.handle, &exitEvent, &resultp)
+	err = processHcsResult(err, resultp)
+	if err != nil {
+		return false, err
+	}
+
+	return waitForSingleObject(exitEvent, timeout)
+}
+
+func (container *container) properties() (*containerProperties, error) {
+	var (
+		resultp     *uint16
+		propertiesp *uint16
+	)
+
+	err := hcsGetComputeSystemProperties(container.handle, "", &propertiesp, &resultp)
+	err = processHcsResult(err, resultp)
+	if err != nil {
+		return nil, err
+	}
+
+	if propertiesp == nil {
+		return nil, errors.New("Unexpected result from hcsGetComputeSystemProperties, properties should never be nil")
+	}
+	propertiesRaw := convertAndFreeCoTaskMemBytes(propertiesp)
+
+	properties := &containerProperties{}
+	if err := json.Unmarshal(propertiesRaw, properties); err != nil {
+		return nil, err
+	}
+
+	return properties, nil
+}
+
+// HasPendingUpdates returns true if the container has updates pending to install
+func (container *container) HasPendingUpdates() (bool, error) {
+	operation := "HasPendingUpdates"
+	title := "HCSShim::Container::" + operation
+	logrus.Debugf(title+" id=%s", container.id)
+	properties, err := container.properties()
+	if err != nil {
+		return false, &ContainerError{Container: container, Operation: operation, Err: err}
+	}
+
+	logrus.Debugf(title+" succeeded id=%s", container.id)
+	return properties.AreUpdatesPending, nil
+}
+
+// Pause pauses the execution of the container. This feature is not enabled in TP5.
+func (container *container) Pause() error {
+	operation := "Pause"
+	title := "HCSShim::Container::" + operation
+	logrus.Debugf(title+" id=%s", container.id)
+	var (
+		//options string
+		resultp        *uint16
+		callbackNumber uintptr
+	)
+
+	if hcsCallbacksSupported {
+		callbackNumber, err := container.registerCallback(hcsNotificationSystemPauseCompleted)
+		if err != nil {
+			return &ContainerError{Container: container, Operation: operation, Err: err}
+		}
+		defer container.unregisterCallback(callbackNumber)
+	}
+
+	err := hcsPauseComputeSystemTP5(container.handle, nil, &resultp)
+	err = processHcsResult(err, resultp)
+	if err != nil {
+		if err == ErrVmcomputeOperationPending {
+			if err := container.waitForNotificationTimeout(callbackNumber, defaultTimeout); err != nil {
+				return &ContainerError{Container: container, Operation: operation, Err: err}
```

Comment:
(and this first part could easily be moved into a helper so that you just have `err = processAsyncHcsResult(err, resultp, callbackNumber, &defaultTimeout)` or something)